source/lac/petsc_precondition.cc

```diff
@@ -1105,18 +1052,100 @@ namespace PETScWrappers
   {
     clear();
 
-    mpi_communicator = matrix_.get_mpi_communicator();
-
-    matrix          = static_cast<Mat>(matrix_);
     additional_data = additional_data_;
 
-    create_pc();
+    create_pc_with_mat(matrix_);
     initialize();
+  }
 
-    PetscErrorCode ierr = PCSetUp(pc);
+  /* ----------------- PreconditionShell -------------------- */
+
+  PreconditionShell::PreconditionShell(const MatrixBase &matrix)
+  {
+    initialize(matrix);
+  }
+
+  PreconditionShell::PreconditionShell(const MPI_Comm &comm)
+  {
+    initialize(comm);
+  }
+
+  void
+  PreconditionShell::initialize(const MPI_Comm &comm)
+  {
+    PetscErrorCode ierr;
+    if (pc)
+      {
+        ierr = PCDestroy(&pc);
+        AssertThrow(ierr == 0, ExcPETScError(ierr));
+      }
+    create_pc_with_comm(comm);
+
+    ierr = PCSetType(pc, PCSHELL);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    ierr = PCShellSetContext(pc, static_cast<void *>(this));
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    ierr = PCShellSetApply(pc, PreconditionShell::pcapply);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    ierr = PCShellSetApplyTranspose(pc, PreconditionShell::pcapply_transpose);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    ierr = PCShellSetName(pc, "deal.II user solve");
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+  }
+
+  void
+  PreconditionShell::initialize(const MatrixBase &matrix)
+  {
+    initialize(matrix.get_mpi_communicator());
+    PetscErrorCode ierr;
+#  if DEAL_II_PETSC_VERSION_LT(3, 5, 0)
+    ierr = PCSetOperators(pc, matrix, matrix, DIFFERENT_NONZERO_PATTERN);
+#  else
+    ierr = PCSetOperators(pc, matrix, matrix);
+#  endif
     AssertThrow(ierr == 0, ExcPETScError(ierr));
   }
 
+  int
+  PreconditionShell::pcapply(PC ppc, Vec x, Vec y)
+  {
+    void *ctx;
+
+    PetscFunctionBeginUser;
+    PetscErrorCode ierr = PCShellGetContext(ppc, &ctx);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    VectorBase src(x);
+    VectorBase dst(y);
+    auto       user = static_cast<PreconditionShell *>(ctx);
+    if (user->apply)
+      user->apply(dst, src);
+    else
+      dst.sadd(0, src);
+    PetscFunctionReturn(0);
+  }
+
+  int
+  PreconditionShell::pcapply_transpose(PC ppc, Vec x, Vec y)
+  {
+    void *ctx;
+
+    PetscFunctionBeginUser;
+    PetscErrorCode ierr = PCShellGetContext(ppc, &ctx);
+    AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+    auto       user = static_cast<PreconditionShell *>(ctx);
+    VectorBase src(x);
+    VectorBase dst(y);
+    if (user->applyT)
+      user->applyT(dst, src);
+    else if (user->apply)
+      user->apply(dst, src);
+    else
+      dst.sadd(0, src);
```

Comment:
Same here. In particular, I don't like the silent fall-back to `apply`. That looks like a hard-to-debug trap.

Reply:
I agree, will add an error here

Reply:
done